### PR TITLE
Payment System Core Repos

### PR DESCRIPTION
**Pull request by stkem**
@andrewconner @leogrim @eishay (who else?)
Next up tests and then Commander logic (currently in a non compiling state, so not in this PR. Also to keep the diff size more manageable)

### CODE REVIEW

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    val clock: Clock) extends DbRepo[PaymentMethod] with PaymentMethodRepo {
+
+  import com.keepit.common.db.slick.DBSession._
+  import db.Driver.simple._
+
+  implicit val stripeTokenColumnType = MappedColumnType.base[StripeToken, String](_.token, StripeToken(_))
+
+  //Put this in a common package? Still needs specific tests
+  //ZZZ NOT CORRECT YET!!
+  implicit val trueOrNullColumnType = MappedColumnType.base[TrueOrNull, Boolean](
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: I spent several hours unsuccessfully trying to make a type mapper that would map `TrueOrNull(true)` to `TRUE` and `TrueOrNull(false)` to `NULL`. I stopped short of implementing my own `JdbcType`. Might be a good use case for our typesafe subscription, @andrewconner ?

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: This is for the same use case as `Keep.primary` isn't it?

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: What about mapping to `Option[Boolean]`?

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: Yup, same use case as `Keep.primary`. If we can get it to work we can switch that one out as well.
Mapping to `Option[Boolean]` unfortunately doesn't work either. Slick treats option types specially (i.e. it's not just an understood type like any other) and that treatment hooks in earlier than the type mappers, so when an `Option` comes out of the type mapper, it doesn't know what to do with it.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala [hunk 2]
```diff
@@ -902,7 +902,7 @@ class LibraryCommanderImpl @Inject() (
     }
   }
 
-  def getLibrariesUserCanKeepTo(userId: Id[User]): Seq[(Library, LibraryMembership, Set[Id[User]])] = { //ZZZ
+  def getLibrariesUserCanKeepTo(userId: Id[User]): Seq[(Library, LibraryMembership, Set[Id[User]])] = {
```
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/LibraryCommander.scala:923]: This is unrelated. Snook in here by accident.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: I'm currently working on adding `primary` to `UserEmailAddress`, anything I should know so that I can switch to your solution when it's ready?

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: Basically just switch out `Option[Boolean]` in your model and table for `TrueOrNull`. Everything else should be taken care of with implicit conversions to and from `Boolean` (i.e. point of use would just use `Boolean` instead of `Option[Boolean]`, which should make that cleaner).

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+case class StripeToken(token: String)
+
+case class PaymentMethod(
+    id: Option[Id[PaymentMethod]] = None,
+    createdAt: DateTime = currentDateTime,
+    updatedAt: DateTime = currentDateTime,
+    state: State[PaymentMethod] = PaymentMethodStates.ACTIVE,
+    accountId: Id[PaidAccount],
+    default: TrueOrNull,
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala]: I'm not a fan of exposing this though, for all purpose this is a `Boolean`. `TrueOrNull` just enables us to have a database constraint, doesn't it? So should we hide this inside of the repo implementation?

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala]: This is the smallest exposure I could come up with. The implicit conversions allow you to just use Booleans at the point of use, i.e. whoever uses the model doesn't really need to know about `TrueOrNull` at all. If you can think of a way to have it just in the repo, without too much Boilerplate, I'm all ears. I'm trying to avoid having to have methods like `applyFromDbRow` and friends just for this.

Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala]: I think @andrewconner mentioned `applyFromDbRow` was actually a good design pattern to have systematically, nope?

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala]: I don't like it very much. Suuuper boilerplaty. You end up effectively spelling out all the fields in you model explicitly _four_ times.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethod.scala]: Agreed, but the time to make a repo is tiny compared to maintaining and dealing with eventual migrations. Without applyFromDb, I can't imagine how hard it would be to migrate models like Keep

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: I've had a nice type mapper for this, on office now but will add it when I get to a place with internet

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    case "user_change_credit" => UserChangeCredit()
+    case "user_added" => extras.as[UserAdded]
+    case "user_removed" => extras.as[UserRemoved]
+    case "admin_added" => extras.as[AdminAdded]
+    case "admin_removed" => extras.as[AdminRemoved]
+    case "plan_changed" => extras.as[PlanChanged]
+    case "payment_method_added" => extras.as[PaymentMethodAdded]
+    case "default_payment_method_changed" => extras.as[DefaultPaymentMethodChanged]
+    case "account_contacts_changed" => extras.as[AccountContactsChanged]
+    case _ => throw new Exception(s"Invalid Event Type: $eventType")
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:123]: I like the strategy we use elsewhere in the codebase:
`case SpecialCredit().eventType => SpecialCredit()`
`case ChargeBack().eventType => ChargeBack()`
but I can see that this has its own merits (for instance, changing the `eventType` string is probably a bad idea because it will lead to issues with existing DB values, and this layout makes that more obvious). I'm not sure where you stand on this.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    state: State[AccountEvent] = AccountEventStates.ACTIVE,
+    eventGroup: Long,
+    eventTime: DateTime,
+    accountId: Id[PaidAccount],
+    billingRelated: Boolean,
+    whoDunnit: Option[Id[User]],
+    whoDunnitExtra: JsValue,
+    kifiAdminInvolved: Option[Id[User]],
+    action: AccountEventAction,
+    creditChange: Int, //in cents
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala]: Do you think it's worth having a type `case class Money(cents: Int)` to make this super explicit everywhere it is used?

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaidPlan.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.joda.time.DateTime
+
+case class PaidPlan(
+    id: Option[Id[PaidPlan]] = None,
+    createdAt: DateTime = currentDateTime,
+    updatedAt: DateTime = currentDateTime,
+    state: State[PaidPlan] = PaidPlanStates.ACTIVE,
+    name: Name[PaidPlan],
+    billingCycle: Int, //in month!
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaidPlan.scala]: Do you think it's worth having a `case class BillingCycle(months: Int)` to make this explicit?

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 4 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:123]: Actually, I see now that in order to use that you'd have to have the `eventType` be a typed object (like `AccountEventActionType`) and copy all this string matching into there. It does introduce a decent amount of boilerplate, but I am a fan of letting the compiler help catch bugs.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 5 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala]: Yes, it probably is. I'm going to change that.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaidPlan.scala (hunk 6 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaidPlan.scala]: Yes :-) See `Money`.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 5 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala]: Even better, it's nice to make a `CurrencyAmount`, which can be constructed using cents and doesn't have a double constructor, but behaves like one.

Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala]: Oh I like that a lot.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala (hunk 1 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: Here's an idea, I haven't compiled this but it should work and may be a bit more semantic?

``` scala
sealed trait PrimaryOrAuxRecord
object PrimaryOrAuxRecord {
  implicit val primaryOrAuxRecordColumnType = MappedColumnType.base[PrimaryOrAuxRecord, Option[Boolean]](toDb, fromDb)

  def fromDb(b: Option[Boolean]) = {
    b match {
      case Some(true) => PrimaryRecord
      case _ => AuxRecord
    }
  }
  def toDb(p: PrimaryOrAuxRecord) = {
    p match {
      case PrimaryRecord => Some(true)
      case AuxRecord => None
    }
  }
}
object PrimaryRecord extends PrimaryOrAuxRecord
object AuxRecord extends PrimaryOrAuxRecord
```

**Comment by andrewconner** [comment]
Get it in! :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+  def toDbRow: (String, JsValue)
+}
+
+object AccountEventAction { //There is probably a deeper type hirachy that can be used here...
+
+  trait Payloadless { self: AccountEventAction =>
+    def toDbRow: (String, JsValue) = (eventType, JsNull)
+  }
+
+  case class SpecialCredit() extends AccountEventAction with Payloadless {
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:25]: `case object`?

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 4 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:123]: Agreed, the repetition could be avoided.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala (hunk 1 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: Just had a chance to finally look at this. I tried this already. I'm afraid you can't have a type mapper that maps to an `Option`, it has to be a primitive type that slick already knows about. See previous comment I made to Léo in this exchange.

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala]: I'm giving up on this for now. Spent way too much time trying to get this to work. @andrewconner maybe ask our typesafe contacts (you know, the ones we are paying for) for an idea some time? Not critical at this point, but I would kinda like to know if there is a good way.

File: kifi-backend/modules/common/conf/evolutions/shoebox/375.sql [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    UNIQUE INDEX payment_method_u_account_id_default (account_id, default)
+);
+
+
+CREATE TABLE paid_account (
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    state varchar(20) NOT NULL,
+    org_id bigint(20) NOT NULL,
```
Comment (leogrim) [inline kifi-backend/modules/common/conf/evolutions/shoebox/375.sql:23]: Have we definitely ruled out personal premium plans?

File: kifi-backend/modules/common/conf/evolutions/shoebox/375.sql [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    CONSTRAINT paid_account_f_organization FOREIGN KEY (org_id) REFERENCES organization(id)
+);
+
+
+CREATE TABLE paid_plan (
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    state varchar(20) NOT NULL,
+    name varchar(256) NOT NULL,
```
Comment (leogrim) [inline kifi-backend/modules/common/conf/evolutions/shoebox/375.sql:39]: Is this `name` a key? Would it be better / more convenient to use it for foreign keys that `id`? It would make other models more readable.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    Some((e.id, e.createdAt, e.updatedAt, e.state, e.eventGroup, e.eventTime, e.accountId,
+      e.billingRelated, e.whoDunnit, e.whoDunnitExtra, e.kifiAdminInvolved, eventType,
+      extras, e.creditChange, e.paymentMethod, e.paymentCharge, e.memo))
+  }
+}
+
+object AccountEventStates extends States[AccountEvent] {
+  val PENDING = State[AccountEvent]("pending")
+  val PROCESSING = State[AccountEvent]("processing")
+  val FAILED = State[AccountEvent]("failed")
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:204]: Do we need the five states? If you do want custom states (I'm not a fan, but let's say sometimes we do), do you need `ACTIVE` and `INACTIVE`? 

Could we stick to either (`ACTIVE`, `INACTIVE`) or a custom set of states? I think problems arise when both are mixed, and `ACTIVE` and `INACTIVE` are no longer reverse of each other.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaidAccountRepo.scala [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+  implicit val dollarAmountColumnType = MappedColumnType.base[DollarAmount, Int](_.cents, DollarAmount(_))
+
+  type RepoImpl = PaidAccountTable
+  class PaidAccountTable(tag: Tag) extends RepoTable[PaidAccount](db, tag, "paid_Account") {
+    def orgId = column[Id[Organization]]("org_id", O.NotNull)
+    def planId = column[Id[PaidPlan]]("plan_id", O.NotNull)
+    def credit = column[DollarAmount]("credit", O.NotNull)
+    def userContacts = column[Seq[Id[User]]]("user_contacts", O.NotNull)
+    def emailContacts = column[Seq[EmailAddress]]("email_contacts", O.NotNull)
```
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaidAccountRepo.scala:33]: What's the benefit of having email addresses in here in addition to users? 
If we do associate several users to a `PaidAccount`, any reason we're denormalizing them as opposed to having another table?

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+  import com.keepit.common.db.slick.DBSession._
+  import db.Driver.simple._
+
+  implicit val stripeTokenColumnType = MappedColumnType.base[StripeToken, String](_.token, StripeToken(_))
+
+  type RepoImpl = PaymentMethodTable
+  class PaymentMethodTable(tag: Tag) extends RepoTable[PaymentMethod](db, tag, "paid_plan") {
+    def accountId = column[Id[PaidAccount]]("account_id", O.NotNull)
+    def default = column[Boolean]("default", O.Nullable)
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala:28]: `column[Option[Boolean]]("default", O.Nullable)` is preferred here (this actually causes problems if you try to do interpolated queries)

File: kifi-backend/modules/common/conf/evolutions/shoebox/375.sql (hunk 8 above)
Comment (andrewconner) [inline kifi-backend/modules/common/conf/evolutions/shoebox/375.sql:23]: I think this is fine, we'll add in personal plans later if we want to. Shouldn't be too insane to refactor if we need to. @stkem?

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 10 above)
Comment (leogrim) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:204]: But really, I think we should rather have a separate column tracking this piece of business logic. `State` should just track whether a row is deleted or not.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala (hunk 12 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala:28]: Wasn't this going to use the `PrimaryOrAuxRecord` (or whatever name you decided)?

Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentMethodRepo.scala:28]: Not reading your email? There was a longer conversation on this pull request (that you were explicitly tagged in). It doesn't work.

File: kifi-backend/modules/common/conf/evolutions/shoebox/375.sql (hunk 8 above)
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/375.sql:23]: Yup, that was kind of the idea behind having a separate model for `PaidAccount` instead of having these things be built into orgs. This model is pretty much entirely contained within the payment system (i.e. it does not appear in any outward facing APIs). It would be quite easy to change this to an `Either[Id[Organization], Id[User]]` (at the scala level, it would be two columns - or possibly another table - in the db)

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaidAccountRepo.scala (hunk 11 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaidAccountRepo.scala:33]: The idea is that you can have non user email addresses as billing contacts (slack supports this), so you can, for example, have bills automatically sent to your (outside) accountant.
Also, this is not a denormalization. The information which contacts are associated with which account is not stored anywhere else.

File: kifi-backend/modules/common/conf/evolutions/shoebox/375.sql (hunk 9 above)
Comment (stkem) [inline kifi-backend/modules/common/conf/evolutions/shoebox/375.sql:39]: Being cleared up on the other PR.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala (hunk 10 above)
Comment (stkem) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/AccountEvent.scala:204]: I disagree with this. These are two different mental models of what state means. To take the example of `PaidPlan`, if we break it out there would be a `state` column (with only `ACTIVE` and `INACTIVE` and another, let’s call it `kind` for now, although I think it’s really a kind of `state`, which would have possible `CUSTOM`, `NORMAL` and `GRANDFATHERED`. Then upon use, you’d have to check `state==`ACTIVE`and`kind ==…`. That makes it more error prone, not less, imho.
